### PR TITLE
#49 - Password reset

### LINKED ISSUE #49: Password reset
- add functionals to rest password

### PR DESCRIPTION
**Pull request by Majkulnn**
- email verification
- password reset

Should close #49

### CODE REVIEW

File: src/Http/Controllers/PasswordResetController.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    public function store(PasswordResetRequest $request): View|RedirectResponse
+    {
+        $status = $this->passwordBrokerManager->sendResetLink(
+            $request->validated(),
+        );
+
+        return $status === PasswordBroker::RESET_LINK_SENT
+            ? \view("user.password.dashboard")->with(["status" => __($status)])
```
Comment (krzysztofrewak) [inline src/Http/Controllers/PasswordResetController.php:37]:
```suggestion
            ? view("user.password.dashboard")->with(["status" => __($status)])
```

Why this `view()` helper is used with slash?

File: src/Http/Controllers/PasswordResetController.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        $status = Password::reset(
+            $request->only("email", "password", "password_confirmation", "token"),
+            function ( $user, $password): void {
+                $user->forceFill([
+                    "password" => $this->hash->make($password),
+                ])->setRememberToken(Str::random(60));
+
+                $user->save();
+
+                event(new PasswordReset($user));
```
Comment (krzysztofrewak) [inline src/Http/Controllers/PasswordResetController.php:57]: Are you sure that this is not used by the framework itself?

File: src/Http/Controllers/Auth/PasswordResetController.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        return view("user.password.forgot-password");
+    }
+
+    public function store(PasswordResetRequest $request): View|RedirectResponse
+    {
+        $status = $this->passwordBrokerManager->sendResetLink(
+            $request->validated(),
+        );
+
+        return $status === PasswordBroker::RESET_LINK_SENT
```
Comment (MrHDOLEK) [inline src/Http/Controllers/Auth/PasswordResetController.php:41]: is huge one liner . Please simplify

File: src/Http/Controllers/Auth/PasswordResetController.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    public function edit(string $token): View
+    {
+        return view("user.password.reset-password")->with("token", $token);
+    }
+
+    public function update(PasswordUpdateRequest $request): RedirectResponse|View
+    {
+        $status = Password::reset(
+            $request->only("email", "password", "password_confirmation", "token"),
+            function ( $user, $password): void {
```
Comment (MrHDOLEK) [inline src/Http/Controllers/Auth/PasswordResetController.php:55]: Not better move to service this function

**Review by MrHDOLEK — CHANGES_REQUESTED** [review]

File: resources/views/user/registered.blade.php [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 @section('content')
     <div style="text-align: center">
-        <h1>You have been registered<br>Please go to login Page</h1>
-        <a href="{{ route('login') }}"><button type="button">Login</button></a>
+        <h1>You have been registered<br>Please verified your email address</h1>
```
Comment (kamilpiech97) [inline resources/views/user/registered.blade.php:5]:
```suggestion
        <h1>You have been registered<br>Please verify your email address</h1>
```

File: src/Http/Controllers/Auth/RegisterController.php [hunk 6]
```diff
@@ -19,6 +20,8 @@ public function create(): View
     public function store(RegisterUserRequest $request): View
     {
         User::create($request->validated());
+        $user = User::query()->where("email", $request->validated("email"))->first();
+        event(new Registered($user));
```
Comment (kamilpiech97) [inline src/Http/Controllers/Auth/RegisterController.php:24]:
```suggestion
        $user = User::create($request->validated());
        event(new Registered($user));
```

File: src/Http/Controllers/Auth/EmailVerificationController.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    public function create(): View
+    {
+        return view("auth.verify-email");
+    }
+
+    public function store(EmailVerificationRequest $request): RedirectResponse
+    {
+        $request->fulfill();
+
+        return redirect("/home");
```
Comment (krzysztofrewak) [inline src/Http/Controllers/Auth/EmailVerificationController.php:24]: Please redirect to named route.

File: src/Http/Controllers/Auth/PasswordResetController.php [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+     * @throws \Illuminate\Auth\AuthenticationException
+     */
+    public function update(PasswordUpdateRequest $request, PasswordResetService $service): RedirectResponse|View
+    {
+        $status = $service->resetPassword($request->validated());
+
+        if ($status === PasswordBroker::PASSWORD_RESET){
+            return view("user.password.dashboard")->with("status", __($status));
+        }
+            return back()->withErrors(["email" => [__($status)]]);
```
Comment (krzysztofrewak) [inline src/Http/Controllers/Auth/PasswordResetController.php:57]: There is something wrong with indentation here.

File: src/Http/Routing/WebRouting.php [hunk 9]
```diff
@@ -19,12 +20,25 @@ public function wire(): void
...
+        $this->router->get("/auth/register", [RegisterController::class, "create"])->name("register");
+        $this->router->post("/auth/register", [RegisterController::class, "store"])->name("register.store");
+        $this->router->get("/auth/login", [LoginController::class, "store"])->name("login");
+        $this->router->post("/auth/login", [LoginController::class, "login"])->name("login.store");
         $this->router->get("/auth/logout", [LoginController::class, "logout"])->name("logout")->middleware("auth");
 
+        $this->router->controller(PasswordResetController::class)->group(function (): void{
+            $this->router->get("/email/verify", "create")->middleware("auth")->name("verification.notice");
+            $this->router->get("/email/verify/{id}/{hash}", "store")->middleware(["auth", "signed"])->name("verification.verify");
+            $this->router->post("/email/verification-notification", "notification")->middleware(["auth", "throttle:6,1"])->name("verification.send");
```
Comment (krzysztofrewak) [inline src/Http/Routing/WebRouting.php:32]: Why `6,1`?

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: src/Http/Routing/WebRouting.php (hunk 9 above)
Comment (MrHDOLEK) [reply src/Http/Routing/WebRouting.php]: @krzysztofrewak It means 6 requests allowed every 1 minute

File: src/Providers/RouteServiceProvider.php [hunk 10]
```diff
@@ -41,5 +41,11 @@ protected function configureRateLimiting(): void
                 $request->user()?->id ?: $request->ip(),
             ),
         );
+        RateLimiter::for(
+            "web",
+            fn(Request $request) => Limit::perMinute(60)->by(
```
Comment (mtracz) [inline src/Providers/RouteServiceProvider.php:46]: Please add return type for arrow function

**Review by mtracz — CHANGES_REQUESTED** [review]

File: resources/views/user/password/forgot-password.blade.php [hunk 11]
```diff
@@ -0,0 +1,20 @@
+@extends('layouts.app')
+
+@section('content')
+    <div>
+        <div>
+            <h1>Forgot Password Page</h1>
```
[the remark this answers is not in the document]
Comment (Majkulnn) [reply resources/views/user/password/forgot-password.blade.php:6]: for now single language,  views right now are just for testing and will be refactored by @JakubZakowicz

File: resources/views/user/password/reset-password.blade.php [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+@section('content')
+    <div>
+        <div>
+            <h1>Forgot Password Page</h1>
+            <form action="{{ route('password.update') }}" method="post">
+                @csrf
+                <div>
+                    <input type="string" id="token" hidden="hidden" name="token" value="{{ $token }}">
+                    <x-input-error for="token"/>
+                </div>
```
[the remark this answers is not in the document]
Comment (Majkulnn) [reply resources/views/user/password/reset-password.blade.php:12]: this token verify the password reset request, is provided from the link clicked by the user in the reset password email

File: src/Http/Controllers/Auth/RegisterController.php [hunk 13]
```diff
@@ -18,7 +19,8 @@ public function create(): View
 
     public function store(RegisterUserRequest $request): View
     {
-        User::create($request->validated());
+        $user = User::query()->create($request->validated());
+        event(new Registered($user));
```
[the remark this answers is not in the document]
Comment (Majkulnn) [reply src/Http/Controllers/Auth/RegisterController.php:23]: it's iluminate/auth/Event and it's not using Dispatchable

**Review by mtracz — DISMISSED** [review]

File: src/Http/Routing/WebRouting.php (hunk 9 above)
Comment (krzysztofrewak) [reply src/Http/Routing/WebRouting.php]: I know what's that. I asked why exactly 6 and 1?

**Review by krzysztofrewak — APPROVED** [review]

**Review by mtracz — APPROVED** [review]

**Review by MrHDOLEK — APPROVED** [review]

**Pull request merged**